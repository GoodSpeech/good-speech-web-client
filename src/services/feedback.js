import { diffWords } from 'diff';
import _ from 'lodash';
import log from 'loglevel';
import { compareTwoTexts } from 'text-sound-similarity';
import words from 'voca/words';


function cleanText(text) {
  return words(text.toLowerCase()).join(' ');
}

function isRemovedPart(part) {
  return part.removed;
}

function isAddedPart(part) {
  return part.added;
}

function isUnchangedPart(part) {
  return !isRemovedPart(part) && !isAddedPart(part);
}

function categorizeAndSequenceDiff(diffed){
  return diffed.reduce((result, part, index) => {
    const [unchanged, added, removed] = result;
    const item = {
      value: part.value,
      position: index
    };

    // The diff library is poorly designed so this check is necessary
    if (isUnchangedPart(part)) {
      unchanged.push({...item, similarity: 1});
    } else if (isAddedPart(part)) {
      added.push(item);
    } else if (isRemovedPart(part)) {
      removed.push(item);
    }
    return result;
  }, [[], [], []]);
}

function getAddedItemsWindow(addedItems, basePoint, size) {
  let start = Math.floor(addedItems.length * basePoint);
  if (start - 1 >= 0) {
    start = start - 1;
  }
  let end = start + size;
  if (end > addedItems.length) {
    end = addedItems.length;
  }
  return addedItems.slice(start, end);
}

function getSimilarity(removedItems, addedItems) {
  return removedItems.map((itemRemoved, index) => {
    const firstWord = cleanText(itemRemoved.value  || '');
    let similarity = 1;
    if (firstWord.length > 1) {
      similarity = getAddedItemsWindow(addedItems, index/removedItems.length, 3).reduce((similarity, itemAdded) => {
        const secondWord = cleanText(itemAdded.value ||Â '');
        let newSimilarity = 0;
        if (secondWord.length >= 1) {
          newSimilarity = compareTwoTexts(firstWord, secondWord);
          log.info(`similarity(${firstWord}, ${secondWord}) = ${newSimilarity}`);
        }
        return _.max([newSimilarity, similarity]);
      }, 0);
    }

    return [itemRemoved, similarity];
  });
}

/**
 * Returns value from the text to read
 *
 * @param  {array} items
 * @return {collection}
 */
function serializeSimilarityArray(items) {
  return items.map((item) => {
    let textToReadItem = item[0];
    return {
      value: textToReadItem.value,
      position: textToReadItem.position,
      similarity: item[1],
      __backup: item
    }
  });
}

function getTextReadedDiff(original, readed) {
  if (readed.length > 0) {
    //this hack improve the diff results
    readed = `${readed.trim()} `;
  }

  const diff = diffWords(original, readed);
  let lastPartMentioned = _.findLastIndex(diff, part => isAddedPart(part) || isUnchangedPart(part));

  const isNextPartPunctuation = diff[lastPartMentioned + 1] && diff[lastPartMentioned + 1].value.length <= 2;
  if (isNextPartPunctuation) {
    lastPartMentioned = lastPartMentioned + 1;
  }

  const lastPartMentionedIsBeforeLast = lastPartMentioned + 1 === diff.length - 1;
  const lastPart = diff[diff.length - 1];

  if (lastPartMentionedIsBeforeLast && lastPart.removed) {
    const lastPartSpacesCount = lastPart.value.split(' ').length;
    if (lastPartSpacesCount <= 4) {
      lastPartMentioned = lastPartMentioned + 1;
    }
  }

  return diff.slice(0, lastPartMentioned + 1);
}

/**
 * @param  {string} original
 * @param  {string} readed
 * @return {array}
 */
function compute(original, readed) {
  let diff, unchanged, added, removed, similarity, merged, serializedSimilarity, sorted;

  diff = getTextReadedDiff(original, readed);
  [unchanged, added, removed] = categorizeAndSequenceDiff(diff);
  similarity = getSimilarity(removed, added);
  serializedSimilarity = serializeSimilarityArray(similarity);
  merged = _.concat(unchanged, serializedSimilarity)
  sorted = _.sortBy(merged, ['position']);

  return sorted;
}

function getScore(computedFeedback) {
  const similarityByWord = computedFeedback.reduce((sentences, sentence) => {
    if (sentence.value !== ' ') {
      const wordsSimilarity = sentence.value.split(' ')
        .filter(word => word !== ' ' && word !== '')
        .map(() => sentence.similarity);
      return sentences.concat(wordsSimilarity);
    }
    return sentences;
  }, [])

  let score = 0;
  if (computedFeedback.length > 0) {
    score = _.sum(similarityByWord) / similarityByWord.length;
  }
  return Math.floor(score * 100);
}

export default {
  compute,
  getScore
};