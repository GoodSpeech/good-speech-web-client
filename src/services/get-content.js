import sample from 'lodash/sample';

import defaultTexts from '../i18n/content';

function getRandomText(lang, isMobile=false) {
    const texts = defaultTexts[lang];
    if (!texts || texts.length === 0) {
        return '';
    }
    if (isMobile) {
        return texts[0];
    }
    return sample(texts);
}

export default getRandomText;
