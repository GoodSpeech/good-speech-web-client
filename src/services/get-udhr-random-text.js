import udhr from 'udhr';
import _ from 'lodash';

const minArticle = 0;
const maxArticle = 9;
const langMapCodes = {
  'cmn-Hans': 'zh-Hans',
  'cmn-Hant': 'zh-Hant',
  'pt': null,
  'de': 'de-1996'
}

function cleanLang(lang) {
  const langCode = Object.keys(langMapCodes).find(langCode => lang.startsWith(langCode))
  if (!langCode) {
    return lang.split('-')[0]
  }

  return langMapCodes[langCode] || lang;
}

function getArticlesAndPreamble(lang) {
  const json = udhr.json();
  const langPrefix = cleanLang(lang)
  const declaration = _.find(json, {lang: langPrefix});
  const preamble = declaration.preamble.para;
  const articles = declaration.article.slice(minArticle, maxArticle).map(article => article.para);
  return [...articles, preamble];
}

function getRandomText(lang) {
  const elements = getArticlesAndPreamble(lang);
  return _.sample(elements);
}

export default getRandomText;