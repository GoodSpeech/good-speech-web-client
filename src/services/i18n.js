/**
 * Translation handler. This is a modified version of:
 * https://jaysoo.ca/2014/03/20/i18n-with-es2015-template-literals/
 */
import translations from '../i18n/translations';

// Matches optional type annotations in i18n strings.
// e.g. i18n`This is a number ${x}:n(2)` formats x as number
//      with two fractional digits.
const typeInfoRegex = /^:([a-z])(\((.+)\))?/;
const config = {
  locale: 'en',
  messageBundle: {},
  _localizers: {
    s /*string*/: v => v.toLocaleString(config.locale),
    n /*number*/: (v, fractionalDigits) => (
      v.toLocaleString(config.locale, {
        minimumFractionDigits: fractionalDigits,
        maximumFractionDigits: fractionalDigits
      })
    )
  }
};

function use(locale) {
  config.locale = locale;
  config.messageBundle = {};
  if (locale in translations) {
    config.messageBundle = translations[locale];
  }
}

function translate(strings, ...values) {
  let translationKey = _buildKey(strings);
  let translationString = config.messageBundle[translationKey] || translationKey;
  let typeInfoForValues = strings.slice(1).map(_extractTypeInfo);
  let localizedValues = values.map((v, i) => _localize(v, typeInfoForValues[i]));

  return _buildMessage(translationString, ...localizedValues);
}

function _extractTypeInfo(str) {
  let match = typeInfoRegex.exec(str);
  if (match) {
    return {type: match[1], options: match[3]};
  } else {
    return {type: 's', options: ''};
  }
}

function _localize(value, {type, options}) {
  return config._localizers[type](value, options);
}

// e.g. _buildKey(['', ' has ', ':c in the']) == '{0} has {1} in the bank'
function _buildKey(strings) {
  let stripType = s => s.replace(typeInfoRegex, '');
  let lastPartialKey = stripType(strings[strings.length - 1]);
  let prependPartialKey = (memo, curr, i) => `${stripType(curr)}{${i}}${memo}`;

  return strings.slice(0, -1).reduceRight(prependPartialKey, lastPartialKey);
}

// e.g. _buildMessage('{0} {1}!', 'hello', 'world') == 'hello world!'
function _buildMessage(str, ...values) {
  return str.replace(/{(\d)}/g, (_, index) => values[Number(index)]);
}


export {translate as i18n, use as useLocale}
