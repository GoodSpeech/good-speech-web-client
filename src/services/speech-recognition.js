import _ from 'lodash';
import hark from 'hark';
import log from 'loglevel';

let speechRecognizer;
let waitingForResultTimeout;
let waitingForActivityTimeout;
let audioStream;
let audioStreamSpeechEvents;
let speaking;

let stopped = false;

const defaultConfig = {
  onRestarting: () => {},
  onRecording: () => {},
  onSpeech: () => {},
  onError: () => {},
  onStart: () => {},
  onStop: () => {},
  lang: 'en-US',
  continuous: true,
  interimResults: true,
  maxAlternatives: 10
};

let currentConfig = defaultConfig;

function clearAllTimeouts() {
  clearTimeout(waitingForResultTimeout);
  clearTimeout(waitingForActivityTimeout);
}

function abortSpeechRecognizer() {
  clearAllTimeouts();
  speechRecognizer.abort();
  log.info('abortSpeechRecognizer');
  currentConfig.onStop();
}

function stopSpeechRecognizer() {
  clearAllTimeouts();
  speechRecognizer.stop();
  log.info('stopSpeechRecognizer');
  currentConfig.onStop();
}

function delayedStartSpeechRecognizer() {
  try {
    _.defer(startSpeechRecognizer);
  } catch (error) {
    delayedStartSpeechRecognizer();
  }
}


function stopSpeakRecognizer() {
  audioStream.getTracks().forEach(track => track.stop());
  audioStreamSpeechEvents.stop();
}

function handleOnEnd() {
  log.info('handleOnEnd');
  clearAllTimeouts();
  currentConfig.onStop();
  log.info('handleOnEnd', 'restarting');
  if (!stopped) {
    currentConfig.onRestarting();
    delayedStartSpeechRecognizer();
  }
}

function handleOnStart() {
  log.info('handleOnStart');
  currentConfig.onRecording();
}

function handleResult(result) {
  if (!result || !result.results) {
    return;
  }

  const recognition = result.results[result.resultIndex];
  const text = _.map(recognition, text => ({
    confidence: text.confidence,
    text: text.transcript
  }));
  log.info('handleResult');
  currentConfig.onSpeech({
    final: recognition.isFinal,
    text
  });
  updateHandleResultTimeout(text, recognition.isFinal);
}

function updateHandleResultTimeout(text, isFinal) {
  clearTimeout(waitingForResultTimeout);
  if (!isFinal) {
    waitingForResultTimeout = setTimeout(() => {
      log.error('result timeout');
      currentConfig.onSpeech({
        final: true,
        text
      });
      abortSpeechRecognizer();
    }, 2000);
  }
}

function handleError(event) {
  log.info('handleError', event.error);
  currentConfig.onError();
}

function handleWaitingForActivityTimeout() {
  if (!speaking) {
    abortSpeechRecognizer();
  }
}

function handleAudioStream(stream) {
  audioStream = stream;
  audioStreamSpeechEvents = hark(audioStream, {
    interval: 200
  });
  audioStreamSpeechEvents.on('speaking', () => {
    speaking = true;
    log.info('speaking');
  });
  audioStreamSpeechEvents.on('stopped_speaking', () => {
    speaking = false;
    clearTimeout(waitingForActivityTimeout);
    waitingForActivityTimeout = setTimeout(handleWaitingForActivityTimeout, 5000);
    log.info('stopped_speaking');
  });
}

function startSpeakRecognizer() {
  navigator.mediaDevices
    .getUserMedia({
      audio: true,
      video: false
    })
    .then(handleAudioStream)
    .catch(err => log.error(err));
}

function startSpeechRecognizer() {
  log.info('startSpeechRecognizer');
  currentConfig.onStart();
  try {
    speechRecognizer.start();
  } catch (err) {
    log.error(err);
  }
}

function initSpeechRecognition({lang, continuous, interimResults, maxAlternatives}) {
  const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition;
  if (SpeechRecognition) {
    speechRecognizer = new SpeechRecognition();
    speechRecognizer.continuous = continuous;
    speechRecognizer.lang = lang;
    speechRecognizer.interimResults = interimResults;
    speechRecognizer.maxAlternatives = maxAlternatives;

    speechRecognizer.onstart = handleOnStart;
    speechRecognizer.onresult = handleResult;
    speechRecognizer.onerror = handleError;
    speechRecognizer.onend = handleOnEnd;  
  }
}

function init(config) {
  const updatedConfig = {
    ...defaultConfig,
    ...config
  };
  currentConfig = updatedConfig;
  initSpeechRecognition(updatedConfig);
}

function start() {
  stopped = false;
  startSpeakRecognizer();
  startSpeechRecognizer();
}

function stop() {
  stopped = true;
  clearAllTimeouts();
  stopSpeakRecognizer();
  stopSpeechRecognizer();
}

export default {
  init,
  start,
  stop
};
