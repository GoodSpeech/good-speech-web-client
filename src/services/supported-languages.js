/*
  You can update the supported language list
  pasting and executing the following code into the console
  of https://cloud.google.com/speech/docs/languages
  
  
  JSON.stringify(
    Array.from($('.devsite-table-wrapper tr'))
      .reduce((languages, tr) => {
        const tds = Array.from(tr.querySelectorAll('td'));
        const texts = tds.map(td => td.innerText);
        if (texts[1]) {
          languages.push({
            name: texts[0],
            code: texts[1],
            englishName: texts[2]
          });
        }
        return languages;
      }, []), null, 2);

*/


/*
  Get the default texts from https://[lang].wikipedia.org/wiki/Wikipedia i.e. https://es.wikipedia.org/wiki/Wikipedia
*/
export const defaultTexts = {
  af: `Wikipedia is 'n veeltalige "kopielinkse" ensiklopedie ontwerp om deur enigiemand gelees en gewysig te word. Dit berus op samewerkende verandering en onderhoud deur duisende gebruikers deur middel van wiki-sagteware, en word verskaf en ondersteun deur die nie-winsgewende Wikimedia Stigting. Waarnemers beskou die tweeledige bestuur, waarin 'n informele gemeenskap van aktiewe gebruikers die gesag oor 'n globale digitale projek met 'n stigting deel en sodoende formele met informele organisasiestrukture verbind word, as 'n eksperiment wat steeds aan die gang is.`,
  id: `Wikipedia adalah proyek ensiklopedia multibahasa dalam jaringan yang bebas dan terbuka, yang dijalankan oleh Wikimedia Foundation, sebuah organisasi nirlaba yang berbasis di Amerika Serikat. Nama Wikipedia berasal dari gabungan kata wiki dan encyclopedia. Wikipedia dirilis pada tahun 2001 oleh Jimmy Wales dan Larry Sanger, dan kini merupakan karya referensi paling besar, cepat berkembang, dan populer di Internet. Proyek Wikipedia bertujuan untuk memberikan ilmu pengetahuan manusia.`,
  ms: `Wikipedia ialah sebuah ensiklopedia yang boleh diubah suai dan dibina secara kerjasama dengan menggunakan perisian wiki. Wikipedia dikendalikan oleh Yayasan Wikimedia yang tidak mengambil apa-apa keuntungan daripada khidmat yang diberikan. Sebagai tambahan kepada matlamat ensiklopedia iaitu untuk memberikan pengetahuan, Wikipedia juga memasukkan maklumat yang lebih kerap dikaitkan dengan almanak dan jurnal, dan juga perkara semasa.`,
  ca: `La Viquipèdia és una enciclopèdia lliure mantinguda per la Fundació Wikimedia, una organització sense ànim de lucre. Els més de 38 milions d'articles han estat escrits col·laborativament per usuaris d'arreu del món, i la majoria dels seus articles poden ser editats per qualsevol persona que pugui accedir a la web.`,
  cs: `Wikipedie je mnohojazyčná webová encyklopedie se svobodným (otevřeným) obsahem, na jejíž tvorbě spolupracují dobrovolní přispěvatelé z celého světa. Jejím cílem je tvorba a celosvětové šíření volně přístupných encyklopedických informací. Wikipedie existuje ve více než 270 jazykových verzích různého rozsahu, přičemž rozsah zhruba třetiny z nich je spíše symbolický.`,
  da: `Wikipedia er en encyklopædi med åbent indhold, skrevet i samarbejde mellem sine brugere. Navnet er en sammentrækning af ordene wiki, der betyder hurtig på hawaiiansk og encyclopedia der betyder encyklopædi på engelsk. Wikipedia styres af Wikimedia, en non-profit fond oprettet specielt til formålet.`,
  de: `Wikipedia ist ein am 15. Januar 2001 gegründetes Projekt zur Erstellung einer Enzyklopädie in zahlreichen Sprachen mit Hilfe des Wiki­prinzips. Es bietet freie, also kostenlose und zur Weiterverbreitung gedachte, unter lexikalischen Einträgen (Lemmata) zu findende Artikel. Das Ziel ist gemäß dem Gründer Jimmy Wales, „eine frei lizenzierte und hochwertige Enzyklopädie zu schaffen und damit lexikalisches Wissen zu verbreiten“.`,
  en: `Wikipedia is a free online encyclopedia with the aim to allow anyone to edit articles. Wikipedia is the largest and most popular general reference work on the Internet and is ranked among the ten most popular websites. Wikipedia is owned by the nonprofit Wikimedia Foundation. Wikipedia was launched on January 15, 2001, by Jimmy Wales and Larry Sanger. Sanger coined its name, a portmanteau of wiki and encyclopedia.`,
  es: `Wikipedia es una enciclopedia libre, políglota y editada de manera colaborativa. Es administrada por la Fundación Wikimedia, una organización sin ánimo de lucro cuya financiación está basada en donaciones. Sus más de 45 millones de artículos en 287 idiomas han sido redactados conjuntamente por voluntarios de todo el mundo,​ lo que hace un total de más de 2000 millones de ediciones, y prácticamente cualquier persona con acceso al proyecto​ puede editarlos, salvo que la página esté protegida contra el vandalismo para evitar problemas y/o trifulcas.`,
  eu: `Wikipedia eduki askeko entziklopedia bat da, lankidetzaz editatua, eleanitza, Interneten argitaratua, Wikimedia Fundazioa irabazi asmorik gabeko erakundeak sustengatua. Wikipedia mundu osoko boluntarioek idazten dute. Internetera konektatutako edonork parte har dezake Wikipediako artikuluetan, aldatu lotura sakatuz. 2015ko azaroaren bostean, 291 hizkuntzatako edizioak zituen, eta horietatik 275 zeuden aktibo.`,
  /*
  fil,
  */
  fr: `Wikipédia Prononciation du titre dans sa version originale Écouter est un projet d'encyclopédie universelle, multilingue, créé par Jimmy Wales et Larry Sanger le 15 janvier 20011 en wiki sous le nom de domaine wikipedia.org. Les versions des différentes langues utilisent le même logiciel de publication, MediaWiki, et ont la même apparence, mais elles comportent des variations dans leurs contenus, leurs structures et leurs modalités d'édition et de gestion.`,
  /*
  gl,
  hr,
  zu,
  is,
  */
  it: `Wikipedia è un'enciclopedia online a contenuto libero, collaborativa, multilingue e gratuita, nata nel 2001, sostenuta e ospitata dalla Wikimedia Foundation, un'organizzazione non a scopo di lucro statunitense. Lanciata da Jimmy Wales e Larry Sanger il 15 gennaio 2001, inizialmente nell'edizione in lingua inglese, nei mesi successivi ha aggiunto edizioni in numerose altre lingue. Sanger ne suggerì il nome, una parola macedonia nata dall'unione della radice wiki al suffisso pedia (da enciclopedia).`,
  /*
  hu,
  nl,
  nb,
  pl,
  */
  pt: `A Wikipédia é um projeto de enciclopédia multilíngue de licença livre, baseado na web e escrito de maneira colaborativa; encontra-se, atualmente, sob administração da Fundação Wikimedia, uma organização sem fins lucrativos cuja missão é "empoderar e engajar pessoas pelo mundo para coletar e desenvolver conteúdo educacional sob uma licença livre ou no domínio público, e para disseminá-lo efetivamente e globalmente.`,
  /*
  ro,
  sk,
  sl,
  fi,
  sv,
  vi,
  tr,
  el,
  bg,
  ru,
  sr,
  uk,
  he,
  ar: 
  fa,
  hi,
  th,
  ko,
  cmn,
  yue,
  */
  ja: `ウィキペディア（英: Wikipedia）は、ウィキメディア財団が運営しているインターネット百科事典である[4]。コピーレフトなライセンスのもと、サイトにアクセス可能な誰もが無料で自由に編集に参加できる[4]。世界の各言語で展開されている。ウィキペディア（Wikipedia）」という名前は、ウェブブラウザ上でウェブページを編集することができる「ウィキ（Wiki）」というシステムを使用した「百科事 であることに由来する造語である。設立者の1人であるラリー・サンガーにより命名された。`
};

export const supportedLanguages = [{
  'name': 'Afrikaans (Suid-Afrika)',
  'code': 'af-ZA',
  'englishName': 'Afrikaans (South Africa)'
}, {
  'name': 'Bahasa Indonesia (Indonesia)',
  'code': 'id-ID',
  'englishName': 'Indonesian (Indonesia)'
}, {
  'name': 'Bahasa Melayu (Malaysia)',
  'code': 'ms-MY',
  'englishName': 'Malay (Malaysia)'
}, {
  'name': 'Català (Espanya)',
  'code': 'ca-ES',
  'englishName': 'Catalan (Spain)'
}, {
  'name': 'Čeština (Česká republika)',
  'code': 'cs-CZ',
  'englishName': 'Czech (Czech Republic)'
}, {
  'name': 'Dansk (Danmark)',
  'code': 'da-DK',
  'englishName': 'Danish (Denmark)'
}, {
  'name': 'Deutsch (Deutschland)',
  'code': 'de-DE',
  'englishName': 'German (Germany)'
}, {
  'name': 'English (Australia)',
  'code': 'en-AU',
  'englishName': 'English (Australia)'
}, {
  'name': 'English (Canada)',
  'code': 'en-CA',
  'englishName': 'English (Canada)'
}, {
  'name': 'English (Great Britain)',
  'code': 'en-GB',
  'englishName': 'English (United Kingdom)'
}, {
  'name': 'English (India)',
  'code': 'en-IN',
  'englishName': 'English (India)'
}, {
  'name': 'English (Ireland)',
  'code': 'en-IE',
  'englishName': 'English (Ireland)'
}, {
  'name': 'English (New Zealand)',
  'code': 'en-NZ',
  'englishName': 'English (New Zealand)'
}, {
  'name': 'English (Philippines)',
  'code': 'en-PH',
  'englishName': 'English (Philippines)'
}, {
  'name': 'English (South Africa)',
  'code': 'en-ZA',
  'englishName': 'English (South Africa)'
}, {
  'name': 'English (United States)',
  'code': 'en-US',
  'englishName': 'English (United States)'
}, {
  'name': 'Español (Argentina)',
  'code': 'es-AR',
  'englishName': 'Spanish (Argentina)'
}, {
  'name': 'Español (Bolivia)',
  'code': 'es-BO',
  'englishName': 'Spanish (Bolivia)'
}, {
  'name': 'Español (Chile)',
  'code': 'es-CL',
  'englishName': 'Spanish (Chile)'
}, {
  'name': 'Español (Colombia)',
  'code': 'es-CO',
  'englishName': 'Spanish (Colombia)'
}, {
  'name': 'Español (Costa Rica)',
  'code': 'es-CR',
  'englishName': 'Spanish (Costa Rica)'
}, {
  'name': 'Español (Ecuador)',
  'code': 'es-EC',
  'englishName': 'Spanish (Ecuador)'
}, {
  'name': 'Español (El Salvador)',
  'code': 'es-SV',
  'englishName': 'Spanish (El Salvador)'
}, {
  'name': 'Español (España)',
  'code': 'es-ES',
  'englishName': 'Spanish (Spain)'
}, {
  'name': 'Español (Estados Unidos)',
  'code': 'es-US',
  'englishName': 'Spanish (United States)'
}, {
  'name': 'Español (Guatemala)',
  'code': 'es-GT',
  'englishName': 'Spanish (Guatemala)'
}, {
  'name': 'Español (Honduras)',
  'code': 'es-HN',
  'englishName': 'Spanish (Honduras)'
}, {
  'name': 'Español (México)',
  'code': 'es-MX',
  'englishName': 'Spanish (Mexico)'
}, {
  'name': 'Español (Nicaragua)',
  'code': 'es-NI',
  'englishName': 'Spanish (Nicaragua)'
}, {
  'name': 'Español (Panamá)',
  'code': 'es-PA',
  'englishName': 'Spanish (Panama)'
}, {
  'name': 'Español (Paraguay)',
  'code': 'es-PY',
  'englishName': 'Spanish (Paraguay)'
}, {
  'name': 'Español (Perú)',
  'code': 'es-PE',
  'englishName': 'Spanish (Peru)'
}, {
  'name': 'Español (Puerto Rico)',
  'code': 'es-PR',
  'englishName': 'Spanish (Puerto Rico)'
}, {
  'name': 'Español (República Dominicana)',
  'code': 'es-DO',
  'englishName': 'Spanish (Dominican Republic)'
}, {
  'name': 'Español (Uruguay)',
  'code': 'es-UY',
  'englishName': 'Spanish (Uruguay)'
}, {
  'name': 'Español (Venezuela)',
  'code': 'es-VE',
  'englishName': 'Spanish (Venezuela)'
}, {
  'name': 'Euskara (Espainia)',
  'code': 'eu-ES',
  'englishName': 'Basque (Spain)'
}, {
  'name': 'Filipino (Pilipinas)',
  'code': 'fil-PH',
  'englishName': 'Filipino (Philippines)'
}, {
  'name': 'Français (Canada)',
  'code': 'fr-CA',
  'englishName': 'French (Canada)'
}, {
  'name': 'Français (France)',
  'code': 'fr-FR',
  'englishName': 'French (France)'
}, {
  'name': 'Galego (España)',
  'code': 'gl-ES',
  'englishName': 'Galician (Spain)'
}, {
  'name': 'Hrvatski (Hrvatska)',
  'code': 'hr-HR',
  'englishName': 'Croatian (Croatia)'
}, {
  'name': 'IsiZulu (Ningizimu Afrika)',
  'code': 'zu-ZA',
  'englishName': 'Zulu (South Africa)'
}, {
  'name': 'Íslenska (Ísland)',
  'code': 'is-IS',
  'englishName': 'Icelandic (Iceland)'
}, {
  'name': 'Italiano (Italia)',
  'code': 'it-IT',
  'englishName': 'Italian (Italy)'
}, {
  'name': 'Lietuvių (Lietuva)',
  'code': 'lt-LT',
  'englishName': 'Lithuanian (Lithuania)'
}, {
  'name': 'Magyar (Magyarország)',
  'code': 'hu-HU',
  'englishName': 'Hungarian (Hungary)'
}, {
  'name': 'Nederlands (Nederland)',
  'code': 'nl-NL',
  'englishName': 'Dutch (Netherlands)'
}, {
  'name': 'Norsk bokmål (Norge)',
  'code': 'nb-NO',
  'englishName': 'Norwegian Bokmål (Norway)'
}, {
  'name': 'Polski (Polska)',
  'code': 'pl-PL',
  'englishName': 'Polish (Poland)'
}, {
  'name': 'Português (Brasil)',
  'code': 'pt-BR',
  'englishName': 'Portuguese (Brazil)'
}, {
  'name': 'Português (Portugal)',
  'code': 'pt-PT',
  'englishName': 'Portuguese (Portugal)'
}, {
  'name': 'Română (România)',
  'code': 'ro-RO',
  'englishName': 'Romanian (Romania)'
}, {
  'name': 'Slovenčina (Slovensko)',
  'code': 'sk-SK',
  'englishName': 'Slovak (Slovakia)'
}, {
  'name': 'Slovenščina (Slovenija)',
  'code': 'sl-SI',
  'englishName': 'Slovenian (Slovenia)'
}, {
  'name': 'Suomi (Suomi)',
  'code': 'fi-FI',
  'englishName': 'Finnish (Finland)'
}, {
  'name': 'Svenska (Sverige)',
  'code': 'sv-SE',
  'englishName': 'Swedish (Sweden)'
}, {
  'name': 'Tiếng Việt (Việt Nam)',
  'code': 'vi-VN',
  'englishName': 'Vietnamese (Vietnam)'
}, {
  'name': 'Türkçe (Türkiye)',
  'code': 'tr-TR',
  'englishName': 'Turkish (Turkey)'
}, {
  'name': 'Ελληνικά (Ελλάδα)',
  'code': 'el-GR',
  'englishName': 'Greek (Greece)'
}, {
  'name': 'Български (България)',
  'code': 'bg-BG',
  'englishName': 'Bulgarian (Bulgaria)'
}, {
  'name': 'Русский (Россия)',
  'code': 'ru-RU',
  'englishName': 'Russian (Russia)'
}, {
  'name': 'Српски (Србија)',
  'code': 'sr-RS',
  'englishName': 'Serbian (Serbia)'
}, {
  'name': 'Українська (Україна)',
  'code': 'uk-UA',
  'englishName': 'Ukrainian (Ukraine)'
}, {
  'name': 'עברית (ישראל)',
  'code': 'he-IL',
  'englishName': 'Hebrew (Israel)'
}, {
  'name': 'العربية (إسرائيل)',
  'code': 'ar-IL',
  'englishName': 'Arabic (Israel)'
}, {
  'name': 'العربية (الأردن)',
  'code': 'ar-JO',
  'englishName': 'Arabic (Jordan)'
}, {
  'name': 'العربية (الإمارات)',
  'code': 'ar-AE',
  'englishName': 'Arabic (United Arab Emirates)'
}, {
  'name': 'العربية (البحرين)',
  'code': 'ar-BH',
  'englishName': 'Arabic (Bahrain)'
}, {
  'name': 'العربية (الجزائر)',
  'code': 'ar-DZ',
  'englishName': 'Arabic (Algeria)'
}, {
  'name': 'العربية (السعودية)',
  'code': 'ar-SA',
  'englishName': 'Arabic (Saudi Arabia)'
}, {
  'name': 'العربية (العراق)',
  'code': 'ar-IQ',
  'englishName': 'Arabic (Iraq)'
}, {
  'name': 'العربية (الكويت)',
  'code': 'ar-KW',
  'englishName': 'Arabic (Kuwait)'
}, {
  'name': 'العربية (المغرب)',
  'code': 'ar-MA',
  'englishName': 'Arabic (Morocco)'
}, {
  'name': 'العربية (تونس)',
  'code': 'ar-TN',
  'englishName': 'Arabic (Tunisia)'
}, {
  'name': 'العربية (عُمان)',
  'code': 'ar-OM',
  'englishName': 'Arabic (Oman)'
}, {
  'name': 'العربية (فلسطين)',
  'code': 'ar-PS',
  'englishName': 'Arabic (State of Palestine)'
}, {
  'name': 'العربية (قطر)',
  'code': 'ar-QA',
  'englishName': 'Arabic (Qatar)'
}, {
  'name': 'العربية (لبنان)',
  'code': 'ar-LB',
  'englishName': 'Arabic (Lebanon)'
}, {
  'name': 'العربية (مصر)',
  'code': 'ar-EG',
  'englishName': 'Arabic (Egypt)'
}, {
  'name': 'فارسی (ایران)',
  'code': 'fa-IR',
  'englishName': 'Persian (Iran)'
}, {
  'name': 'हिन्दी (भारत)',
  'code': 'hi-IN',
  'englishName': 'Hindi (India)'
}, {
  'name': 'ไทย (ประเทศไทย)',
  'code': 'th-TH',
  'englishName': 'Thai (Thailand)'
}, {
  'name': '한국어 (대한민국)',
  'code': 'ko-KR',
  'englishName': 'Korean (South Korea)'
}, {
  'name': '國語 (台灣)',
  'code': 'cmn-Hant-TW',
  'englishName': 'Chinese, Mandarin (Traditional, Taiwan)'
}, {
  'name': '廣東話 (香港)',
  'code': 'yue-Hant-HK',
  'englishName': 'Chinese, Cantonese (Traditional, Hong Kong)'
}, {
  'name': '日本語（日本）',
  'code': 'ja-JP',
  'englishName': 'Japanese (Japan)'
}, {
  'name': '普通話 (香港)',
  'code': 'cmn-Hans-HK',
  'englishName': 'Chinese, Mandarin (Simplified, Hong Kong)'
}, {
  'name': '普通话 (中国大陆)',
  'code': 'cmn-Hans-CN',
  'englishName': 'Chinese, Mandarin (Simplified, China)'
}]

export default supportedLanguages;