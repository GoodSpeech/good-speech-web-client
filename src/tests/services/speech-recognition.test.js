import * as SpeechRecognition from '../../services/speech-recognition';

describe('SpeechRecognition', () => {
  const tmpSpeechRecognition = window.SpeechRecognition;
  const startMock = jest.fn();
  const stopMock = jest.fn();

  beforeEach(() => {
    window.SpeechRecognition = function() {
      this.start = startMock;
      this.stop = stopMock;
    };
  });

  afterEach(() => {
    window.SpeechRecognition = tmpSpeechRecognition;
  });

  describe('Init', () => {
    it('should initialize the SpeechRecognition service properly', () => {
      expect(SpeechRecognition.init({})).toMatchSnapshot();
    });
  });

  describe('Start', () => {
    it('should start the SpeechRecognition service properly', () => {
      const onStart = jest.fn();
      SpeechRecognition.init({onStart});
      SpeechRecognition.start();
      expect(startMock.mock.calls.length).toBe(1);
      expect(onStart.mock.calls.length).toBe(1);
    });
  });

  describe('Stop', () => {
    it('should stop the SpeechRecognition service properly', () => {
      const onStop = jest.fn();
      SpeechRecognition.init({
        onStart: () => {},
        onStop
      });
      SpeechRecognition.start();
      SpeechRecognition.stop();
      expect(stopMock.mock.calls.length).toBe(1);
      expect(onStop.mock.calls.length).toBe(1);
    });
  });
});