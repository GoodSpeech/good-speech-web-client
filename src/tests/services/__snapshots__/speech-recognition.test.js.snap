// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeechRecognition Init should initialize the SpeechRecognition service properly 1`] = `
 {
  "continuous": true,
  "interimResults": true,
  "lang": "en-US",
  "maxAlternatives": 10,
  "onend": [Function],
  "onerror": [Function],
  "onresult": [Function],
  "onstart": [Function],
  "start": [Function],
  "stop": [Function],
}
`;
